lib/api/patchable_api.rb

```diff
@@ -0,0 +1,71 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+module API
+  module PatchableAPI
+    def self.included(base)
+      base.extend ClassMethods
+    end
+
+    module ClassMethods
+      def inherited(subclass)
+        super
+
+        # run unscoped patches (i.e. patches that are on the class root, not in a namespace)
+        subclass.send(:execute_patches_for, nil)
+      end
+
+      def namespace(name, *args, &block)
+        super(name, *args) do
+          instance_eval(&block)
+          execute_patches_for(name)
+        end
+      end
+
+      # we need to repeat all the aliases for them to work properly...
+      alias_method :group, :namespace
+      alias_method :resource, :namespace
+      alias_method :resources, :namespace
+      alias_method :segment, :namespace
```

Comment:
Okay, so `alias_method_chain` will not work, because the original `alias_method` calls all created their own **copy** of the `namespace` method. That is exactly the behaviour we would be using in an `alias_method_chain`. Thus our own redefinition and `alias_method_chain`-ing would have no effect on those original `alias_method`s.

Quick example:
````ruby
[1] pry(main)> class Klaus
[1] pry(main)*   def foo  
[1] pry(main)*     'lame'    
[1] pry(main)*   end  
[1] pry(main)*   alias_method :bar, :foo
[1] pry(main)*   alias_method :foo_without_coolness, :foo  
[1] pry(main)*   def foo_with_coolness
[1] pry(main)*     "cool, not #{foo_without_coolness}"
[1] pry(main)*   end  
[1] pry(main)*   alias_method :foo, :foo_with_coolness
[1] pry(main)* end  
=> Klaus
[2] pry(main)> Klaus.new.foo
=> "cool, not lame"
[3] pry(main)> Klaus.new.bar
=> "lame"
````